cherry-pick: fix out-of-bounds array deref caught by asan

dassaf4 opened the pull request:
Cherry-pick of #92 

Details:
* Longstanding oob error caused by incomplete buffer -> bvector conversion
* ConvexHullXY closure point logic appended last point instead of first

EarlinLutz reviewed (changes requested):
This code was probably ported from old-style pre-allocated arrays that were certain to have sufficient size.

I think xyz1.push_back (xyz1[0]) still has potential problems if the size and capacity are already equal -- xyz1[0] should be copied into a local DPoint3d before being pushed.  But I'll be delighted if push_back is smart enough to not deallocate the input-time memory block before copying to a newly allocated block.

dassaf4 commented:
> I think xyz1.push_back (xyz1[0]) still has potential problems if the size and capacity are already equal...

Great point! I changed the push_back to use a local copy of the point, and also added a test to confirm that the compiler appends a valid point on push_back realloc just in case, as I'm sure I've seen (and used!) the anti-pattern elsewhere.

MichaelBelousov commented:
latest changes still pass the reported repro
